<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Александр! Давай знакомиться! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на "ты" 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).


Вижу, что ты очень старался, делая проект. Все ключевые этапы в работе выполнены, и в целом с проектом ты справился. 
    
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Можно учесть эти комментарии при выполнении будущих заданий или доработать текущий проект (однако это не обязательно). 

<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы "со звездочкой", которые помогут в будущем.

    
Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».
</div>

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble 

In [2]:
data = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
display (data.head())
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Данные загружены и исследованы.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Рекомендую разделить код на несколько частей, каждую выполнить в своей ячейке. Отдельно выполнить импорты, отдельно загрузку данных, отдельно исследование. Так проект будет читаться легче.

 <div class="alert alert-info"> <b>Комментарий студента: </b> Сделал</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
<b> v2 </b>
Обрати внимание, что не выведен результат вызова data.head(). Или оберни команду в display() или вызови её в последней строке ячейки.  
</div>

## Разбейте данные на выборки

### Разобьем данные сначала на тренировочную и проверочную выборки. 

In [4]:
features = data.drop('is_ultra', axis=1)
target = data['is_ultra']
features_train, features_sum, target_train, target_sum= train_test_split(features, target, test_size=0.4, random_state=12345)


### Проверочную выборку в свою очередь разобьем на валидационную и тестовую выборки

In [5]:
features_val, features_test, target_val, target_test = train_test_split(features_sum, target_sum, test_size=0.5, random_state=12345)

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, корректно разделил данные на три выборки.
На будущее, обрати внимание на аргумент stratify функции train_test_split. Стратификация особенно важна, когда у нас мало данных и возможен дисбаланс классов в обучении / валидации / тесте.
</div>

## Исследуйте модели

### Модель дерева решений

In [6]:
%%time
best_accuracy = 0
best_depth = 0
for depth in range(1,10):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_train, target_train)
    accuracy_train = model.score(features_train, target_train)
    accuracy_val = model.score(features_val, target_val)
    #predictions=model.predict(features_val) мой вопрос по этой строке ниже 
    #accuracy_1=accuracy_score (target_val, predictions) мой вопрос по этой строке ниже 
    if accuracy_val > best_accuracy:
        best_accuracy = accuracy_val
        best_depth = depth
    print("Глубина дерева:", depth, "    Качество тренировочной выборки:", accuracy_train)    
    print("Глубина дерева:", depth, "    Качество валидационной выборки:", best_accuracy)
    #print("Глубина дерева:", depth, "    Качество валидационной выборки:",  accuracy_1)
print("Лучшие параметры:      Глубина дерева:", best_depth, "    Качество:", best_accuracy)




Глубина дерева: 1     Качество тренировочной выборки: 0.7577800829875518
Глубина дерева: 1     Качество валидационной выборки: 0.7542768273716952
Глубина дерева: 2     Качество тренировочной выборки: 0.7878630705394191
Глубина дерева: 2     Качество валидационной выборки: 0.7822706065318819
Глубина дерева: 3     Качество тренировочной выборки: 0.8075726141078838
Глубина дерева: 3     Качество валидационной выборки: 0.7853810264385692
Глубина дерева: 4     Качество тренировочной выборки: 0.8106846473029046
Глубина дерева: 4     Качество валидационной выборки: 0.7853810264385692
Глубина дерева: 5     Качество тренировочной выборки: 0.8200207468879668
Глубина дерева: 5     Качество валидационной выборки: 0.7853810264385692
Глубина дерева: 6     Качество тренировочной выборки: 0.8376556016597511
Глубина дерева: 6     Качество валидационной выборки: 0.7853810264385692
Глубина дерева: 7     Качество тренировочной выборки: 0.8558091286307054
Глубина дерева: 7     Качество валидационной выборк

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично:
    
 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Лучше использовать больше 5 значений гиперпараметров.
 <div class="alert alert-info"> <b>Комментарий студента: </b> Сделал</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Рекомендую посмотреть как при изменении гиперпараметров меняется точность на обучающем и валидационых наборах. Для этого с помощью print() выводи эти метрики на каждом шагу цикла.
 <div class="alert alert-info"> <b>Комментарий студента: </b> Сделал + добавил вывод. Поясни пожалуйста, в чем разница между accuracy_val и  accuracy_1 ??: accuracy_val = model.score(features_val, target_val)    и  predictions=model.predict(features_val); accuracy_1=accuracy_score (target_val, predictions). Это одинаковые функции, только вторая требует импорта библиотеки и выполняется в 2 этапа?  Видимо я не до конца разобрался</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 Это два разных способа получит один и тот же результат.</div>

<div class="alert alert-info">
<font color='black'> <b> При глубине дерева "3" качество на валидационной выборке достигло максимума и уже не меняется </font> <b>

### Модель логистической регрессии

In [7]:
%%time
model = LogisticRegression()
model.fit(features_train, target_train)
accuracy = model.score(features_val, target_val)
print("Качество:", accuracy)

Качество: 0.7107309486780715
CPU times: user 19 ms, sys: 3.5 ms, total: 22.5 ms
Wall time: 21.1 ms


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Логистическая регрессия исследована
</div>


### Модель случайного леса

In [8]:
%%time
best_accuracy = 0
best_est = 0
for est in range(1,40):
    model = RandomForestClassifier(n_estimators=est, random_state=12345)
    model.fit(features_train, target_train)
    accuracy_train = model.score(features_train, target_train)
    accuracy_val = model.score(features_val, target_val)
    if accuracy_val > best_accuracy:
        best_accuracy = accuracy_val
        best_est = est
    print("Количество деревьев:", est, "    Качество тренировочной выборки:", accuracy_train)    
    print("Количество деревьев:", est, "    Качество валидационной выборки:", best_accuracy)
print("Лучшие параметры:     Количество деревьев:", best_est, "    Качество:", best_accuracy)


Количество деревьев: 1     Качество тренировочной выборки: 0.9035269709543569
Количество деревьев: 1     Качество валидационной выборки: 0.7107309486780715
Количество деревьев: 2     Качество тренировочной выборки: 0.9097510373443983
Количество деревьев: 2     Качество валидационной выборки: 0.7636080870917574
Количество деревьев: 3     Качество тренировочной выборки: 0.9507261410788381
Количество деревьев: 3     Качество валидационной выборки: 0.7636080870917574
Количество деревьев: 4     Качество тренировочной выборки: 0.946058091286307
Количество деревьев: 4     Качество валидационной выборки: 0.7713841368584758
Количество деревьев: 5     Качество тренировочной выборки: 0.9678423236514523
Количество деревьев: 5     Качество валидационной выборки: 0.7713841368584758
Количество деревьев: 6     Качество тренировочной выборки: 0.9693983402489627
Количество деревьев: 6     Качество валидационной выборки: 0.7807153965785381
Количество деревьев: 7     Качество тренировочной выборки: 0.9787

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично:
    
 - исследовано более 5 значений гиперпараметров
 - модель обучена на обучающем наборе
 - получена оценка качества на валидационном наборе
 - перебор гиперпараметров осуществляется в цикле
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Рекомендую посмотреть как при изменении гиперпараметров меняется точность на обучающем и валидационых наборах. Для этого с помощью print() выводи эти метрики на каждом шагу цикла.
 <div class="alert alert-info"> <b>Комментарий студента: </b> Сделал + добавил вывод. </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Рассмотрены все три изученных алгоритма (решающее дерево, случайный лес, логистическая регрессия)
</div>


<div class="alert alert-info">
<font color='black'> <b>На валидационной выборке наилучшим образом показала себя модель случайного леса: качество составило 0.794. Видно, что при числе деревьев=40 , точность на обучающей выборке почти идеальная<b>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец, есть выводы, и выбрана лучшая модель, включая гиперпараметры!
</div>


## Проверьте модель на тестовой выборке

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
К данному этапу мы должны подойти с одной выбранной моделью. Любое использование тестовой выборки (в том числе для выбора лучшей модели) - это нарушение. <div class="alert alert-info"> <b>Комментарий студента: </b> Не совсем понял, что надо сделать. Модель я выбрал, просто проверил все три на тестовой выборке. У всех произошло снижение accuracy </div>

<div class="alert alert-info"> 
<b>v2</b>


Попробую пояснить логику. Давай представим себе, что у нас есть бизнес-задача - сделать модель, которая должна рекомендовать тарифный план. Заказчика не интересуют наши заморочки - выборки, модели, количество деревьев... Ему  нужна модель, которая даст хорошее качество решения.
    
Для себя мы ставим целью сделать это качество получше, рассмотрели три модели, подобрали гиперпараметры. Теперь мы должны выбрать ОДНУ модель, которая и будет работать у заказчика. Три модели ему не нужны.  Теперь вопрос: по какой выборке мы выбираем эту лучшую модель. Моя рекомендация - делать это по валидационной выборке, тестовая выборка на выбор модели влиять не должна. А  когда мы захотим оценить точность нашей модели на данных, которые она не видела, мы должны сделать это на тестовой выборке.
    
Теперь как это соотносится с тем, что ты провел тестирование трех моделей. Если ты сделал это просто из научного любопытства - ОК, нет проблем. Проблемы могут возникнуть, если ты вдруг решишь выбрать в качестве лучшей модели ту, которая оказалась лучшей по результатам тестирования, а не по результатам валидации.  Как только ты испльзуешь информацию о точности на тестовой выборке для выбора модели ты совершаешь ошибку.
 
</div>

### Модель дерева решений

In [9]:
model = DecisionTreeClassifier(max_depth=3, random_state=12345)
model.fit(features_train, target_train)
accuracy = model.score(features_test, target_test)
display("Качество:", accuracy)

'Качество:'

0.7791601866251944

### Модель логистической регрессии

In [10]:
model = LogisticRegression()
model.fit(features_train, target_train)
accuracy = model.score(features_test, target_test)
display("Качество:", accuracy)

'Качество:'

0.6842923794712286

### Модель случайного леса

In [11]:
model = RandomForestClassifier(n_estimators=23, random_state=12345)
model.fit(features_train, target_train)
accuracy = model.score(features_test, target_test)
display("Качество:", accuracy)

'Качество:'

0.7807153965785381

<div class="alert alert-info">
<font color='black'> <b>Проверим лучшую модель на совмещенной выборке (валидационная + обучающая). Из-за большего набора данных модель обучилась лучше и качество соответственно увеличилось и составило 0.8 <b>

In [12]:
model = RandomForestClassifier(n_estimators=23, random_state=12345)
feat = pd.concat([features_train, features_val], ignore_index=True)
target = pd.concat([target_train, target_val], ignore_index=True)
model.fit(feat, target)
accuracy1 = model.score(features_test, target_test)
display("Качество:", accuracy1)

'Качество:'

0.8055987558320373

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Молодец: 
    
  - выбрана лучшая модель
  - тестирование произведено корректно 
  - достигнута хорошая точность на тестовой выборке
</div>


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Кстати, есть честный способ повысить точность модели путем увеличения размера выборки. На этапе тестирования валидационная выборка уже выполнила свою задачу - мы подобрали гиперпарамеры. Теперь мы можем обучить нашу модель на увеличенном наборе данных (валидация + обучение) и протестировать на тестовом наборе.<div class="alert alert-info"> <b>Комментарий студента: </b> Сделал </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>v2</b> 👍 </div>

## (бонус) Проверьте модели на адекватность

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [ ] Рассмотрено больше одной модели
    - [ ] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [ ] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75


<div class="alert alert-info">
<font color='black'> <b>ВЫВОД: На тестовой выборке наилучшим образом показала себя снова модель случайного леса: качество тем не менее немного ухудшилось и составило 0.78

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>

В целом мне все понравилось: твоя работа выполнена на хорошем уровне с минимальными помарками.
Также работа выглядит аккуратной и хорошо оформленной.

К твоей работе у меня нет серьезных замечаний и проект может быть принят. Возвращаю проект чтобы дать тебе возможность ознакомиться с моими комментариями, возможно у тебя возникнут вопрсы или ты решишь внести изменения в проект.
    
Жду твою работу на финальное ревью :)
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий v2</b>

Привет. Я постарался ответить на все твои вопросы. 
Проект принят.  Поздравляю и желаю дальнейших успехов!
</div>